### ISSUE TITLE
LogMessageWaitStrategy for DockerComposeContainer

### ISSUE THREAD

**Issue by s2929346** [testcontainers/testcontainers-java#515]
Hi,

I am starting some docker containers via a `docker-compose.yml` file. In code, I use the `DockerComposeContainer` class as following:

```
DockerComposeContainer c = new DockerComposeContainer(<path>).withTailChildContainers(true);
c.starting(null);
```

I would like to use the `LogMessageWaitStrategy` (https://github.com/testcontainers/testcontainers-java/blob/master/core/src/main/java/org/testcontainers/containers/wait/LogMessageWaitStrategy.java) on this container. I must wait until certain log events occurred before I can start using the containers. The `LogMessageWaitStrategy` should be perfect for this.

Is this possible? I am having a hard time to figure out how I should do this.

Thanks guys!

**Comment by s2929346** [comment]
Hi,

Update from my side. I got it working with the following code:

```
        DockerClient client = DockerClientFactory.instance().client();
        WaitingConsumer consumer = new WaitingConsumer();

        LogUtils.followOutput(client, <id of container>, consumer, OutputFrame.OutputType.STDOUT, OutputFrame.OutputType.STDERR);

        try {
            consumer.waitUntil(frame -> frame.getUtf8String().matches(<regex>), 300, TimeUnit.SECONDS, 1);
        } catch (TimeoutException e) {
            // todo - handle exception properly
        } finally {
	    // todo - unregister logger/consumer!
	}
```

The only problem left here is that I should unregister the consumer in the `finally` clause. If I don't do this, I think I will run out of memory eventually, because `WaitingConsumer.frames` is basically unbounded. That said, I have no idea how to do this. I could execute `DockerClientFactory.instance().client().stop()`, but that will probably kill the whole test-containers library.

I could try to initialize my own docker client, instead of using `DockerClientFactory.instance().client()`. Do you have any other ideas?

Are there plans to integrate the wait strategy-functionality on `DockerComposeContainer`?

Thanks guys!

**Comment by rnorth** [comment]
Hi @s2929346 
As you've seen I'm afraid we don't have the log wait functionality for docker-compose yet. I can't see it being _too_ difficult to add, but it might still not be something we could add quickly due to other work that's ongoing. Sorry.

What you're doing looks fine as a workaround. As you point out `WaitingConsumer.frames` being unbounded could cause a problem if your logs are that big, so that is a bug that I'd like to address (as this would affect all users, not just people using docker-compose). We could probably drop log frames as they're consumed, which would be significantly more efficient. Will have a look.
Thanks
Richard

**Comment by s2929346** [comment]
Hi @rnorth,

Thank you for your thoughts here. I'll continue using my current implementation for now then. An in-house support for this would be great though!

With regards to `WaitingConsumer` and unbounded buffer: I think you can remove a logging frame once it has been evaluated by the predicate. Since the regex is matched against each frame separately, we don't need the frame anymore after predicate evaluation. And now, looking again to the actual code in `WaitingConsumer`, this is exactly what you are doing in `frames.pollLast(100, TimeUnit.MILLISECONDS)`. So I think there is no real problem after all. My apologies for the confusion.

While writing this down, it made me wonder if a `multi-line regex wait` on the complete log would ever be useful. This is currently not supported, as each frame is checked individually in `WaitingConsumer`. This would also make the logic above more complex, as you need to keep the frames longer and need to decide when to remove them anyway.

Thanks!

**Comment by s2929346** [comment]
Right, I remember what I was thinking yesterday: It might be useful to support that a `WaitingConsumer` (and its underlying docker container callback) could be removed/disposed somehow. I think a `WaitingConsumer`'s job becomes obsolete once the waiting predicate has evaluated to true (of the timeout has reached). Now, after the `waitUntil` returns or throws the time-out exception, the `WaitingConsumer` is still receiving frames. And this will lead to the out of memory eventually. There is no way to stop this.

**Comment by bsideup** [comment]
Fixed by #600 (to be released very soon)

**Comment by bsideup** [comment]
Also, Duplicates #174. Will close this one.
